File: src/java/com/thoughtworks/webdriver/firefox/FirefoxWebElement.java
package com.thoughtworks.webdriver.firefox;

import java.util.ArrayList;
import java.util.List;

import com.thoughtworks.webdriver.WebElement;

public class FirefoxWebElement implements WebElement {
	private final String elementId;
	private final ExtensionListener listener;

	public FirefoxWebElement(ExtensionListener listener, String elementId) {
		this.listener = listener;
		this.elementId = elementId;
	}

	public void click() {
		listener.writeAndWaitForResponse("click", elementId);
	}

	public String getAttribute(String name) {
		return null;
	}

	public List getChildrenOfType(String tagName) {
		String childIds = listener.writeAndWaitForResponse("getChildrenOfType", elementId + " " + tagName);
		String[] ids = childIds.split(",");
		List children = new ArrayList();
		for (int i = 0; i < ids.length; i++) 
			children.add(new FirefoxWebElement(listener, ids[i]));
		return children;
	}

	public String getText() {
		return listener.writeAndWaitForResponse("getElementText", elementId);
	}

	public String getValue() {
		return listener.writeAndWaitForResponse("getElementValue", elementId);
	}

	public void setValue(String value) {
		listener.writeAndWaitForResponse("setElementValue", elementId + " " + value);
	}
	
	public boolean isEnabled() {
		return false;
	}

	public boolean isSelected() {
		return false;
	}

	public void setSelected() {
	}

	public void submit() {
	}

	public boolean toggle() {
		return false;
	}	
}

File: bindings/py/src/com/googlecode/webdriver/firefox/FirefoxWebElement.py
from ExtensionConnection import ExtensionConnection
from common import logger
from com.googlecode.webdriver.WebElement import WebElement

class FirefoxWebElement(WebElement):

  def __init__(self, parent, id):
    self.parent = parent
    self.conn = ExtensionConnection()
    self.id = id

  def GetText(self):
    return self._command("getElementText")

  def Click(self):
    self._command("click")

  def Submit(self):
    self._command("submitElement")

  def GetValue(self):
    return self._command("getElementValue")

  def Clear(self):
    self._command("clear")

  def GetAttribute(self, name):
    return self._command("getElementAttribute", name)

  def Toggle(self):
    self._command("toggleElement")

  def IsSelected(self):
    return self._command("getElementSelected")

  def SetSelected(self):
    self._command("setElementSelected")

  def IsEnabled(self):
    if self.GetAttribute("disabled"):
      return False
    else:
      # The "disabled" attribute may not exist
      return True

  def FindElements(self, by):
    pass

  def FindElement(self, by):
    pass

  def FindElementsByXPath(self, xpath):
    resp = self._command("findElementsByXPath", xpath)
    elems = []
    for elemId in resp.split(","):
      elem = FirefoxWebElement(self.parent, elemId)
      elems.append(elem)
    return elems

  def _command(self, _cmd, *args):
    return self.conn.command(_cmd, params = args, elementId=self.id)
